### PR TITLE
Fix `decimal type` Input Validation Error

### PR DESCRIPTION
**Pull request by DimuthuMadushan**
## Purpose

Fixes: [#2649](https://github.com/ballerina-platform/ballerina-standard-library/issues/2649)

## Checklist
- [x] Linked to an issue
- [x] Updated the changelog
- [x] Added tests

### CODE REVIEW

File: ballerina/validator_visitor.bal [hunk 1]
```diff
@@ -234,8 +235,8 @@ class ValidatorVisitor {
         if value is () {
             if schemaArg.'type.kind == NON_NULL {
                 string listError = getListElementError(self.argumentPath);
-                string message = string`${listError}Expected value of type "${getTypeNameFromType(schemaArg.'type)}"` +
-                string`, found null.`;
+                string message = string `${listError}Expected value of type "${getTypeNameFromType(schemaArg.'type)}"` +
+                string `, found null.`;
```
Comment (ThisaruGuruge) [inline ballerina/validator_visitor.bal:239]: More readable this way. Shall we check the other places as well?

```suggestion
                string message = string `${listError}Expected value of type "${getTypeNameFromType(schemaArg.'type)}"` +
                                 string `, found null.`;
```

File: ballerina/validator_visitor.bal [hunk 2]
```diff
@@ -276,21 +277,17 @@ class ValidatorVisitor {
...
                         } else {
                             string expectedTypeName = getOfType(subInputValue.'type).name.toString();
-                            if expectedTypeName == FLOAT && argValue is int {
-                                variableValues[subInputValue.name] = <float> argValue;
-                            }
-                            self.validateArgumentValue(argValue, location, getTypeNameFromValue(fieldValue),
-                                                       subInputValue);
+                            string actualTypeName = getTypeNameFromValue(fieldValue);
+                            variableValues[subInputValue.name] = self.coerceValue(fieldValue, expectedTypeName,
+                                actualTypeName, location);
```
Comment (ThisaruGuruge) [inline ballerina/validator_visitor.bal:287]:
```suggestion
                            variableValues[subInputValue.name] = self.coerceValue(fieldValue, expectedTypeName,
                                                                                  actualTypeName, location);
```

File: ballerina/validator_visitor.bal [hunk 3]
```diff
@@ -326,33 +323,19 @@ class ValidatorVisitor {
                 return;
             }
             if variableValues.length() > 0 {
-                foreach int i in 0..< variableValues.length() {
+                foreach int i in 0 ..< variableValues.length() {
```
Comment (ThisaruGuruge) [inline ballerina/validator_visitor.bal:326]: Ditto

File: ballerina/validator_visitor.bal [hunk 4]
```diff
@@ -326,33 +323,19 @@ class ValidatorVisitor {
...
-                        } else {
-                            string message = string`${expectedTypeName} cannot represent non ${expectedTypeName} `+
-                            string`value: ${listItemValue.toString()}`;
-                            ErrorDetail errorDetail = getErrorDetailRecord(message, location);
-                            self.errors.push(errorDetail);
+                            string actualTypeName = getTypeNameFromValue(listItemValue);
+                            variableValues[i] = self.coerceValue(listItemValue, expectedTypeName, actualTypeName,
+                                location);
+                            self.validateArgumentValue(listItemValue, location, getTypeNameFromValue(listItemValue),
+                                listItemInputValue);
```
Comment (ThisaruGuruge) [inline ballerina/validator_visitor.bal:338]: Check formatting.

File: ballerina/validator_visitor.bal [hunk 5]
```diff
@@ -367,25 +350,62 @@ class ValidatorVisitor {
...
+            float|error coerceValue = float:fromString(value.toString());
+            if coerceValue is float {
+                return coerceValue;
+            } else {
+                string message = string `${expectedTypeName} cannot represent non ${expectedTypeName} value: ` +
+                string `${value.toString()}`;
+                ErrorDetail errorDetail = getErrorDetailRecord(message, location);
+                self.errors.push(errorDetail);
+            }
+        } else if expectedTypeName == DECIMAL && (actualTypeName == INT || actualTypeName == FLOAT) {
```
Comment (ThisaruGuruge) [inline ballerina/validator_visitor.bal:397]: From both `if` and `else` blocks, `actualTypeName == INT` in common. Is there a way to take it out? That will reduce the complexity and increase the readability of the code.

Or maybe we can simplify the logic using boolean transformations.

File: changelog.md [hunk 6]
```diff
@@ -11,6 +11,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 ### Fixed
 - [[#2640] Fix inter-dependent fragments returning stack overflow error](https://github.com/ballerina-platform/ballerina-standard-library/issues/2640)
+- [[#2649] Fix invalid validation of `decimal` type values](https://github.com/ballerina-platform/ballerina-standard-library/issues/2649)
```
Comment (ThisaruGuruge) [inline changelog.md:14]:
```suggestion
- [[#2649] Fix incorrectly invalidating `decimal` type inputs](https://github.com/ballerina-platform/ballerina-standard-library/issues/2649)
```

File: ballerina/validator_visitor.bal (hunk 5 above)
Comment (DimuthuMadushan) [reply ballerina/validator_visitor.bal:397]: Taking `actualTypeName == INT` out  may increase the size of the function. Any suggestion for a simplified logic? or we can add an inner `if` block like 
`else if expectedTypeName == DECIMAL {
            if actualTypeName == INT || actualTypeName == FLOAT {`

File: ballerina/validator_visitor.bal [hunk 7]
```diff
@@ -367,27 +350,68 @@ class ValidatorVisitor {
...
+    isolated function coerceValue(Scalar value, string expectedTypeName, string actualTypeName, Location location)
+        returns Scalar {
+        if expectedTypeName == FLOAT {
+            if actualTypeName == INT || actualTypeName == DECIMAL {
+                float|error coerceValue = float:fromString(value.toString());
+                if coerceValue is float {
+                    return coerceValue;
+                } else {
+                    string message = string `${expectedTypeName} cannot represent non ${expectedTypeName} value: ` +
+                                     string `${value.toString()}`;
```
Comment (ThisaruGuruge) [inline ballerina/validator_visitor.bal:394]: Do we need to call the `toString()` here?

File: ballerina/validator_visitor.bal [hunk 8]
```diff
@@ -555,16 +560,16 @@ class ValidatorVisitor {
                         __InputValue? inputValue = getInputValueFromArray(defaultDirective.args, argName);
                         if inputValue is __InputValue {
                             _ = notFoundInputValues.remove(<int>notFoundInputValues.indexOf(inputValue));
-                            self.visitArgument(argumentNode, {input:inputValue, fieldName:directive.getName()});
+                            self.visitArgument(argumentNode, {input: inputValue, fieldName: directive.getName()});
                         } else {
-                            string message = string`Unknown argument "${argName}" on directive` +
-                            string`"${directive.getName()}".`;
+                            string message = string `Unknown argument "${argName}" on directive` +
+                            string `"${directive.getName()}".`;
```
Comment (ThisaruGuruge) [inline ballerina/validator_visitor.bal:566]: Formatting?

**Review by ThisaruGuruge — APPROVED** [review]

File: ballerina/validator_visitor.bal (hunk 5 above)
Comment (ThisaruGuruge) [reply ballerina/validator_visitor.bal]: This looks good.

**Pull request merged**